calico/election.py

```diff
@@ -0,0 +1,213 @@
+
+"""
+calico.election
+~~~~~~~~~~~~
+
+Calico election code.
+"""
+import etcd
+import eventlet
+from httplib import HTTPException
+import logging
+from socket import timeout as SocketTimeout
+import time
+from urllib3 import Timeout
+from urllib3.exceptions import ReadTimeoutError, ConnectTimeoutError, HTTPError
+
+import os
+
+_log = logging.getLogger(__name__)
+
+
+class ElectionReconnect(Exception):
+    """
+    Exception indicating that an error occurred, and we should try to
+    reconnect.
+    """
+    pass
+
+
+class Elector(object):
+    def __init__(self, client, server_id, election_key,
+                 interval=30, ttl=60):
+        """
+        Class that manages elections.
+
+        :param client: etcd client object
+        :param server_id: Server ID. Must be unique to this server, and should
+                          take a value that is meaningful in logs (e.g.
+                          hostname)
+        :param election_key: The etcd key used in the election - e.g.
+                             "/calico/v1/election"
+        :param interval: Interval (seconds) between checks on etcd. Must be > 0
+        :param ttl: Time to live (seconds) for etcd values. Must be > interval.
+        """
+        self._etcd_client = client
+        self._server_id = server_id
+        self._key = election_key
+        self._interval = int(interval)
+        self._ttl = int(ttl)
+
+        if self._interval <= 0:
+            raise ValueError("Interval %r is <= 0" % interval)
+
+        if self._ttl <= self._interval:
+            raise ValueError("TTL %r is <= interval %r" % (ttl, interval))
+
+        # Used to track whether or not this is the first iteration of the
+        # attempts to test who is master.
+        self._first_iteration = True
+
+        # Is this the master? To start with, no
+        self._master = False
+
+        # Keep the greenlet ID handy to ease UT.
+        self._greenlet = eventlet.spawn(self._run)
+
+    def _run(self):
+        """
+        Main election thread run routine.
+
+        The slightly artificial split between this and _vote is mostly so that
+        we can easily catch and log any exception that takes out the greenlet.
+        """
+        try:
+            while True:
+                try:
+                    self._vote()
+                except ElectionReconnect:
+                    # Something failed, and wants us just to go back to the
+                    # beginning.
+                    pass
+
+        except:
+            # Election greenlet failed. Log but reraise.
+            _log.exception("Election greenlet exiting")
+            raise
+
+    def _vote(self):
+        """
+        Main election thread routine to reconnect and perform election.
+        """
+        # Sleep only if this is not the first time round the loop.
+        if self._first_iteration:
+            self._first_iteration = False
+        else:
+            eventlet.sleep(self._interval)
```

Comment:
You can skip this weirdness by bringing the sleep into the loop in `_run`: that'll save you an instance variable and this if block.